### ISSUE TITLE
Verify compliant metric SDK specification implementation: MetricReader/MetricReader operations/RegisterProducer(metricProducer)

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go#3661]
- [ ] Identify all the normative requirements, recommendations, and options [the specification] defines as comments to this issue
- [ ] Ensure the [current metric SDK implementation] is compliant with these normative requirements, recommendations, and options in those comments.

[the specification]: https://github.com/open-telemetry/opentelemetry-specification/blob/0c5bfa999a5310801d5c36d2a82cbcd0e32ca9cb/specification/metrics/sdk.md#registerproducermetricproducer
[current metric SDK implementation]: https://github.com/open-telemetry/opentelemetry-go/tree/v1.12.0/sdk/metric

**Comment by damemi** [comment]
Requirements from the [MetricReader RegisterProducer(metricProducer) section](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/sdk.md#registerproducermetricproducer):

> * RegisterProducer causes the MetricReader to use the provided [MetricProducer](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/sdk.md#metricproducer) as a source of aggregated metric data in subsequent invocations of Collect. RegisterProducer is expected to be called during initialization, but MAY be invoked later.

> * Multiple registrations of the same MetricProducer MAY result in duplicate metric data being collected.

> * If the [MeterProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/sdk.md#meterprovider) is an instance of [MetricProducer](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/sdk.md#metricproducer), this MAY be used to register the MeterProvider, but MUST NOT allow multiple [MeterProviders](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/sdk.md#meterprovider) to be registered with the same MetricReader.

---

The SDK implements [`periodicReader.RegisterProducer()`](https://github.com/open-telemetry/opentelemetry-go/blob/c404a30b96f6117ac677c5e15119a4d724e2ab3e/sdk/metric/periodic_reader.go#L198) and [`manualReader.RegisterProducer()`](https://github.com/open-telemetry/opentelemetry-go/blob/c404a30b96f6117ac677c5e15119a4d724e2ab3e/sdk/metric/manual_reader.go#L71). Both of which can be called at initialization or after.

There is nothing preventing multiple registrations of the same Producer. New registrations are simply appended to the in-memory list of Producers. In `Collect()`, all Producers will be called (including duplicates).

On the last point, [`MeterProvider`](https://github.com/open-telemetry/opentelemetry-go/blob/c404a30b96f6117ac677c5e15119a4d724e2ab3e/sdk/metric/provider.go#L32) in the SDK does not implement `MetricProducer`, but technically I think an SDK author could write a MeterProvider that does implement MetricProducer. In that case, our SDK's `RegisterProducer()` functions don't check if the `metricProducer` is a MeterProvider. This way, it's possible to register multiple MeterProviders to the same MetricReader (though we ensure only 1 is treated like a MeterProvider, they are all collected). We could enforce this stricter by checking if the producer implements a MeterProvider, and if so make sure that no sdkProducer is registered.

**Comment by MrAlias** [comment]
> On the last point, [`MeterProvider`](https://github.com/open-telemetry/opentelemetry-go/blob/c404a30b96f6117ac677c5e15119a4d724e2ab3e/sdk/metric/provider.go#L32) in the SDK does not implement `MetricProducer`, but technically I think an SDK author could write a MeterProvider that does implement MetricProducer. In that case, our SDK's `RegisterProducer()` functions don't check if the `metricProducer` is a MeterProvider. This way, it's possible to register multiple MeterProviders to the same MetricReader (though we ensure only 1 is treated like a MeterProvider, they are all collected). We could enforce this stricter by checking if the producer implements a MeterProvider, and if so make sure that no sdkProducer is registered.

Yeah, I think the specification is a bit unclear on this, but I think that given our `MeterProvider` does not implement the `MetricProducer` interface the requirement that multiple of them cannot be registered does not apply. If a user decides to wrap our `MeterProvider` and create their own `MeterProvider` that does implement a `MetricProducer` it would follow that their implementation would need to make sure the restriction is followed.

I think we are compliant with the specification.

**Comment by MrAlias** [comment]
Closing as this looks done.
